tracing: replace most tracing infrastructure with a custom tracer

RaduBerinde opened the pull request:
We are jumping through many hoops to coax the basictracer to do what we need:
recorders, span event listeners, tee tracer, nested tee traces, swapping tracers
on the go. We would need more hacks to make things work nicely with
gRPC-opentracing interceptors (#9806).

This change replaces most of the tracing infrastructure (including TeeTracer)
with a custom Tracer which supports:

  - forwarding events to x/net/trace instances

  - recording traces. Recording is started automatically for spans that have
    the Snowball baggage and can be started explicitly as well. Recorded
    events can be retrieved at any time.

  - lightstep traces. This is implemented by maintaining a "shadow" lightstep
    span inside each of our spans.

When using lightstep, the tracer relies on the lightstep contexts, and extracts
trace and span IDs from them.

A single Tracer is used throughout a server (we no longer need to create
temporary tracers for snowball traces). The Tracer is used even when tracing is
disabled (for its recording capabilities); in this case we don't create spans
(unless needed for recording) and callers can still circumvent expensive Event
calls.

The main purpose of the change is to open the door for cleanly
implementing #9806, but it also fixes two problems as a bonus:

  - EXPLAIN (TRACE) now works with sql.trace.txn.enable_threshold (#15193)

  - when using lightstep with EXPLAIN (TRACE) or sql.trace.txn.enable_threshold,
    the parent/child relationship with remote spans is now maintained correctly
    (this wasn't the case before, because the lightstep trace/span IDs were
    different from the recorded ones).

We now also have the capability to get records of non-Finish()ed spans.

We still rely on basictracer for a few limited things which we may clean up in
the future: context injection/extraction, RawSpan (we probably want to use a
proto for collected spans instead of gob encoding/decoding RawSpans).

Checked a profile under kv read-only load with tracing disabled and made some
optimizations; tracing functions are now insignificant.

Fixes #15193.

RaduBerinde committed:
tracing: replace most tracing infrastructure with a custom tracer

We are jumping through many hoops to coax the basictracer to do what we need:
recorders, span event listeners, tee tracer, nested tee traces, swapping tracers
on the go. We would need more hacks to make things work nicely with
gRPC-opentracing interceptors (#9806).

This change replaces most of the tracing infrastructure (including TeeTracer)
with a custom Tracer which supports:

  - forwarding events to x/net/trace instances

  - recording traces. Recording is started automatically for spans that have
    the Snowball baggage and can be started explicitly as well. Recorded
    events can be retrieved at any time.

  - lightstep traces. This is implemented by maintaining a "shadow" lightstep
    span inside each of our spans.

When using lightstep, the tracer relies on the lightstep contexts, and extracts
trace and span IDs from them.

A single Tracer is used throughout a server (we no longer need to create
temporary tracers for snowball traces). The Tracer is used even when tracing is
disabled (for its recording capabilities); in this case we don't create spans
(unless needed for recording) and callers can still circumvent expensive Event
calls.

The main purpose of the change is to open the door for cleanly
implementing #9806, but it also fixes two problems as a bonus:

  - EXPLAIN (TRACE) now works with sql.trace.txn.enable_threshold (#15193)

  - when using lightstep with EXPLAIN (TRACE) or sql.trace.txn.enable_threshold,
    the parent/child relationship with remote spans is now maintained correctly
    (this wasn't the case before, because the lightstep trace/span IDs were
    different from the recorded ones).

We now also have the capability to get records of non-Finish()ed spans.

We still rely on basictracer for a few limited things which we may clean up in
the future: context injection/extraction, RawSpan (we probably want to use a
proto for collected spans instead of gob encoding/decoding RawSpans).

Checked a profile under kv read-only load with tracing disabled and made some
optimizations; tracing functions are now insignificant.

Fixes #15193.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16058)
<!-- Reviewable:end -->

tamird commented:
Reviewed 12 of 14 files at r1, 1 of 1 files at r2.
Review status: 13 of 14 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/server/node.go, line 913 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkgRNgOMgWhACS0IKk:-KkkgRNgOMgWhACS0IKl:b-4htuxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/server/node.go#L913)):*
> ```Go
> 	var sp opentracing.Span
> 	remoteParent := false
> 	if parentSp := opentracing.SpanFromContext(ctx); parentSp != nil {
> ```

nit: `parentSp` is never used (except for comparing to nil)

---

*[pkg/server/node.go, line 937 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkgfhPYMRD3JQcZFcC:-KkkgfhPYMRD3JQcZFcD:bvtf0pv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/server/node.go#L937)):*
> ```Go
> 
> 	finishSpan := func(br *roachpb.BatchResponse) {
> 		sp.Finish()
> ```

this (and others like it) is an unrelated refactor, right? just avoiding an extra lookup into the context?

---

*[pkg/server/node.go, line 941 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkgFsTN51hRt-_d_Ez:-KkkgFsTN51hRt-_d_F-:b-4lhqk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/server/node.go#L941)):*
> ```Go
> 			return
> 		}
> 		if remoteParent || true {
> ```

hm, this is suspicious

---

*[pkg/server/node.go, line 944 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkgmcnnzc1DNiKc3Ol:-Kkkgmcnnzc1DNiKc3Om:b5l8k50) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/server/node.go#L944)):*
> ```Go
> 			// If this is a "snowball trace", we'll need to encode all the recorded
> 			// spans in the BatchResponse at the end of the request.
> 			if r := tracing.GetRecording(sp); r != nil {
> ```

nit: I don't think this nil check is needed, you can just `range` over it directly.

---

*[pkg/sql/trace.go, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-VVe3NF2bsyEg9V1:-Kkl-VVe3NF2bsyEg9V2:b2i7tc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L168)):*
> ```Go
> 		for i := 0; i < len(recording); i++ {
> 			if len(recording[i].Logs) == 0 {
> 				copy(recording[i:], recording[i+1:])
> ```

won't this skip examining any entry that follows an empty entry?

---

*[pkg/sql/trace.go, line 174 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkruJDcBQj2YGnKAMy:-KkkruJETMknem9yq8YM:bh5ygnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L174)):*
> ```Go
> 		if len(recording) == 0 {
> 			if !n.exhausted {
> 				fmt.Printf("\n\n ADDING EMPTY SPAN TO RECORD \n\n")
> ```

debugging?

---

*[pkg/sql/trace.go, line 182 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-ksxeLgU7XxYgl9O:-Kkl-ksxeLgU7XxYgl9P:b-hryruf) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L182)):*
> ```Go
> 				}
> 				recording = tracing.GetRecording(n.recordingSp)
> 				fmt.Printf("\n\n RECORDING: %+v\n\n", recording)
> ```

debugging

---

*[pkg/sql/trace.go, line 231 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-C5VcapFLeEa88RX:-Kkl-C5VcapFLeEa88RY:b2fd4sa) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L231)):*
> ```Go
> 			}
> 		}
> 		// Clear the spans that have been accumulated so far, so that we'll
> ```

this comment is incorrect now, right?

---

*[pkg/sql/distsqlrun/base.go, line 610 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkpwT4JQNLeNJewr2A:-KkkpwT4JQNLeNJewr2B:bid215z) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/distsqlrun/base.go#L610)):*
> ```Go
> 	req.TraceContext = &tracing.SpanContextCarrier{}
> 	err := tracer.Inject(sp.Context(), basictracer.Delegator, req.TraceContext)
> 	// Clear the trace context if it wasn't initialized.
> ```

nit: would't this be clearer as:
```
var traceContext tracing.SpanContextCarrier
if err := tracer.Inject(sp.Context(), basictracer.Delegator, &traceContext); err != nil {
  return err
}
req.TraceContext = &traceContext
return nil
```
?

The difference is the `err == nil && TraceID == 0` case; can that happen?

---

*[pkg/util/tracing/tracer.go, line 173 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl0HfIQ6e8ScCO5OPd:-Kkl0HfIQ6e8ScCO5OPe:bcwvvjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer.go#L173)):*
> ```Go
> 
> 	if t.lightstep != nil {
> 		var opts []opentracing.StartSpanOption
> ```

slightly scary that removing this line would still compile but would do the wrong thing

---

*[pkg/util/tracing/tracer_span.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkj1BsQkvXsNm8f0De:-Kkkj1BsQkvXsNm8f0Df:b-ezrzy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L56)):*
> ```Go
> }
> 
> var _ opentracing.SpanContext = spanContext{}
> ```

just wondering: why implement the interface on the value rather than pointer?

---

*[pkg/util/tracing/tracer_span.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkjEqI6sMH9-bhcddv:-KkkjEqI6sMH9-bhcddw:bs8ppjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L81)):*
> ```Go
> 
> 	// Atomic flag used to avoid taking the mutex in the hot path.
> 	recording int32
> ```

nit: newline after?

---

*[pkg/util/tracing/tracer_span.go, line 78 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-zZLzvq_yY5nrM3r:-Kkl-zZLzvq_yY5nrM3s:b-d81xsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L78)):*
> ```Go
> 	operation string
> 	startTime time.Time
> 	duration  time.Duration
> ```

i think this one is unused (except for a place where it is always set to -1)

---

*[pkg/util/tracing/tracer_span.go, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkoom6L0KHW-DkdM5N:-Kkkoom6L0KHW-DkdM5O:b-ott0w3) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L168)):*
> ```Go
> 	s.mu.Unlock()
> 	if group == nil {
> 		return errors.Errorf("adding Raw Spans to a span that isn't recording")
> ```

errors.New?

---

*[pkg/util/tracing/tracer_span.go, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkotCrFj2fcvOSRQVV:-KkkotCrFj2fcvOSRQVW:b-dw6x9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L177)):*
> ```Go
> 
> // ClearRecordedLogs removes all logs from the recording.
> // TODO(radu): this is temporary; only used because of
> ```

might be good to expand on this. concretely, i have no idea what this means.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058)*
<!-- Sent from Reviewable.io -->

tamird commented:
Excited to see work in this direction! Someone more knowledgable (probably @andreimatei) should look at this as well, of course.

---

Review status: 13 of 14 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-:-Kkl1VFG9WjIlgaNN8DA:b-anhl81)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM  Thanks for doing this.

I'd (more emphatically) name another major motivation in the commit message - getting access to non-finished recorded spans.

---

Review status: 13 of 14 files reviewed at latest revision, 28 unresolved discussions, some commit checks failed.

---

*[pkg/util/tracing/tracer.go, line 17 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkkkhvED6acV62COP-:-KkkkkhvED6acV62COP0:b-e790lr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L17)):*
> ```Go
> // Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
> // Author: Andrei Matei (andreimatei1@gmail.com)
> // Author: Radu Berinde (radu@cockroachlabs.com)
> ```

third time will be a charm with this file!

---

*[pkg/util/tracing/tracer.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkl-ANcV8s8zJq32wS:-Kkkl-ANcV8s8zJq32wT:b5jrww0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L53)):*
> ```Go
> //    events can be retrieved at any time.
> //
> //  - lightstep traces. This is implemented by maintaining a "shadow" lightstep
> ```

is anything here Lightstep-specific, or did you mean to say "opentracing"?
I think it's a matter of time until we include a 2nd opentracing tracer in crdb (someone was already asking about Zipkin).

---

*[pkg/util/tracing/tracer.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkklOlkY_ewQy1nCMS9:-KkklOlkY_ewQy1nCMSA:bv8tojk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L63)):*
> ```Go
> 	netTrace bool
> 	// If set, we are using a lightstep tracer for most operations.
> 	lightstep opentracing.Tracer
> ```

ditto making it more generic. `wrappedTracer` perhaps.

---

*[pkg/util/tracing/tracer.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkklwnQYYKKC_XKlUoF:-KkklwnQYYKKC_XKlUoG:bwgktjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L104)):*
> ```Go
> //
> // When tracing is disabled all spans are noopSpans; these spans aren't
> // capable of recording, so if StartRecording is to be used, this option should
> ```

nit: s/if StartRec is to be used/if the caller wants to reserve the right to call StartRec
to suggest that the caller might not know if it'll want to record in the future.

---

*[pkg/util/tracing/tracer.go, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkknIaSC5baqnrnUlUW:-KkknIaSC5baqnrnUlUX:bbkmk03) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L127)):*
> ```Go
> 	force := false
> 
> 	for _, o := range opts {
> ```

this function is pretty long. How about some comments along the way?

Here:
```
// Decode the options.
```

---

*[pkg/util/tracing/tracer.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkknkOAzH9hC2DIl0xY:-KkknkOAzH9hC2DIl0xZ:bdfh1sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L172)):*
> ```Go
> 	s.duration = -1
> 
> 	if t.lightstep != nil {
> ```

`// Create the shadow lightstep/opentracing span.`

---

*[pkg/util/tracing/tracer.go, line 221 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl4TEXbbdX7t0lZplh:-Kkl4TEYDQa1q6msAmF9:bl8ba4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer.go#L221)):*
> ```Go
> 	}
> 
> 	if hasParent {
> ```

can we unify this block with the other block above that calls `enableRecording`? So that the conditions for enabling recording are all visible in one place.

Also, it bothers me a little that a span now has two things that control recording now - the "snowball" baggage item and the `recordingGroup != nil`. Could we not use the baggage item any more, except when injecting the trace (so, for propagating the recording information across the network)? We'd terminate the baggage item in `JoinRemoteTrace`.

---

*[pkg/util/tracing/tracer_span.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkko6jJGC3-4bjfKxXv:-Kkko6jJGC3-4bjfKxXw:be9rpyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L74)):*
> ```Go
> 	tracer    *Tracer
> 	netTr     trace.Trace
> 	lightstep opentracing.Span
> ```

if this can be nil, please comment

---

*[pkg/util/tracing/tracer_span.go, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkomp7XNpNug8r9eju:-Kkkomp7XNpNug8r9ejv:brpg4b6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L89)):*
> ```Go
> 		recordedLogs   []opentracing.LogRecord
> 		// tags are only set if we are recording.
> 		tags opentracing.Tags
> ```

but what if recording is started after the span is created (and presumably tags are usually set on the span when the span is created)?

---

*[pkg/util/tracing/tracer_span.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkpFlbGAZo5M82WQQ7:-KkkpFlbGAZo5M82WQQ8:b-5w0y3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L160)):*
> ```Go
> }
> 
> // AddRawSpansToRecord adds raw span data to the record. Used to import recorded
> ```

nit: "to the record" is not very clear... Perhaps `ImportRemoteSpans()`? The `spanGroup` already talks about "spans from other hosts"

---

*[pkg/util/tracing/tracer_span.go, line 354 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkq3UgJNmQD1mn59sy:-Kkkq3UgJNmQD1mn59sz:b-axfg5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L354)):*
> ```Go
> type spanGroup struct {
> 	syncutil.Mutex
> 	spans []*span
> ```

I think it'd be useful to keep track of which of the `spans/rawSpans` are finished, if we could get this information. I can add this later too, so I'd like to understand what's possible. I see that for `spans`, some of them are finished, some of them are not. And currently all the `rawSpans` are finished remote spans, right? But they don't have to be, if we'd have a way to ask remotes for unfinished spans.

---

*[pkg/util/tracing/tracer_span.go, line 78 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-6gXkDBLN2Urq__C:-Kkl-6gYQ6jHaQlrw3eg:b-tlkj3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L78)):*
> ```Go
> 	operation string
> 	startTime time.Time
> 	duration  time.Duration
> ```

I think this is set on `Finish()`. If so, consider adding that as a comment. Particularly since this field is used to tell if a span has been finished or not.

---

*[pkg/util/tracing/tracer_span.go, line 263 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl8hrSBIx7Vy_zac-B:-Kkl8hrSBIx7Vy_zac-C:b-k0euoe) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L263)):*
> ```Go
> }
> 
> // LogFIelds is part of the opentracing.Span interface.
> ```

Log*Fi*elds

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-:-KkkoUxLoJTRA2To2e8D:b-uevg3x)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on pkg/server/node.go:
```diff
@@ -908,44 +908,48 @@ func (n *Node) setupSpanForIncomingRPC(
-				} else {
-					// We can't log to the finished span; strip the span from the context.
-					log.Warning(opentracing.ContextWithSpan(ctx, nil), err)
+		if remoteParent || true {
```
Did you mean to trip this always?

tschottdorf commented on pkg/sql/trace.go:
```diff
@@ -159,26 +160,41 @@ func (n *explainTraceNode) Next(ctx context.Context) (bool, error) {
+		if len(recording) == 0 {
 			if !n.exhausted {
-				n.trace.AddDummySpan(basictracer.RawSpan{
+				fmt.Printf("\n\n ADDING EMPTY SPAN TO RECORD \n\n")
```
Remove this line.

tschottdorf commented on pkg/sql/trace.go:
```diff
@@ -159,26 +160,41 @@ func (n *explainTraceNode) Next(ctx context.Context) (bool, error) {
+					return false, err
+				}
+				recording = tracing.GetRecording(n.recordingSp)
+				fmt.Printf("\n\n RECORDING: %+v\n\n", recording)
```
Ditto.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (Haven't reviewed this as thoroughly as I would've liked due to the size, but I don't think there's a good way to split it up either). Thanks for tackling this!

---

Reviewed 13 of 14 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 31 unresolved discussions, some commit checks failed.

---

*[pkg/sql/trace.go, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-VVe3NF2bsyEg9V1:-KklESdR2WBzW3Af723G:b-rdy6m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L168)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

won't this skip examining any entry that follows an empty entry?
</blockquote></details>

Yeah this seems off-by-one - good catch!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-:-KklUJYOLxlIAKPim-Oa:b-wz59vq)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks for the reviews!

---

Review status: all files reviewed at latest revision, 31 unresolved discussions, some commit checks failed.

---

*[pkg/server/node.go, line 913 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkgRNgOMgWhACS0IKk:-KklNjaDEP9eLb71kgZo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/server/node.go#L913)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: `parentSp` is never used (except for comparing to nil)
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 937 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkgfhPYMRD3JQcZFcC:-KklNkQpOXA8JP8gYRYU:b-nrakw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/server/node.go#L937)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this (and others like it) is an unrelated refactor, right? just avoiding an extra lookup into the context?
</blockquote></details>

Yeah, I noticed it on the profile.

---

*[pkg/server/node.go, line 941 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkgFsTN51hRt-_d_Ez:-KklNqTKWyMtTv9YlkXn:bgm022u) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/server/node.go#L941)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

hm, this is suspicious
</blockquote></details>

Good catch, thanks guys! I was debugging something and forgot that.

---

*[pkg/server/node.go, line 944 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkgmcnnzc1DNiKc3Ol:-KklO7gF69Qnix4uIKzB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/server/node.go#L944)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: I don't think this nil check is needed, you can just `range` over it directly.
</blockquote></details>

Done.

---

*[pkg/sql/trace.go, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-VVe3NF2bsyEg9V1:-KklOA7CCdIe9teb1-QD:b-7qy51i) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L168)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

won't this skip examining any entry that follows an empty entry?
</blockquote></details>

Great catch, thanks!! Fixed.

---

*[pkg/sql/trace.go, line 174 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkruJDcBQj2YGnKAMy:-KklOPxQgXkXELOKQ0Ht:b32gium) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L174)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

debugging?
</blockquote></details>

Removed, sorry

---

*[pkg/sql/trace.go, line 182 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-ksxeLgU7XxYgl9O:-KklORYt8eEBHYE4uBsZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L182)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

debugging
</blockquote></details>

Done.

---

*[pkg/sql/trace.go, line 231 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-C5VcapFLeEa88RX:-KklO_rzyjLszwXYUJW5:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/trace.go#L231)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this comment is incorrect now, right?
</blockquote></details>

Fixed.

---

*[pkg/sql/distsqlrun/base.go, line 610 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkpwT4JQNLeNJewr2A:-KklOm0z4dEoPybzuwBC:bkmlzd9) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/sql/distsqlrun/base.go#L610)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: would't this be clearer as:
```
var traceContext tracing.SpanContextCarrier
if err := tracer.Inject(sp.Context(), basictracer.Delegator, &traceContext); err != nil {
  return err
}
req.TraceContext = &traceContext
return nil
```
?

The difference is the `err == nil && TraceID == 0` case; can that happen?
</blockquote></details>

It can happen if it's a noop span. But we are actually checking that above. Done.

---

*[pkg/util/tracing/tracer.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkl-ANcV8s8zJq32wS:-KklPGpxXOKdoEiJX4jK:bk0uk21) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L53)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is anything here Lightstep-specific, or did you mean to say "opentracing"?
I think it's a matter of time until we include a 2nd opentracing tracer in crdb (someone was already asking about Zipkin).
</blockquote></details>

it is specific - we are relying on it using basictracer underneath, to be able to retrieve TraceID/SpanIDs. This actually won't be the case anymore when we update lightstep, and this code will need some minor changes to adjust.

That said, it should be easy to add others, we will likely need a slightly richer superset of opentracing.Tracer.

---

*[pkg/util/tracing/tracer.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkklwnQYYKKC_XKlUoF:-KklQ0YoB_LUMxjiRkZh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L104)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: s/if StartRec is to be used/if the caller wants to reserve the right to call StartRec
to suggest that the caller might not know if it'll want to record in the future.
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkknkOAzH9hC2DIl0xY:-KklSuD0pkZXVKJ5OuHg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer.go#L172)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`// Create the shadow lightstep/opentracing span.`
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer.go, line 173 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl0HfIQ6e8ScCO5OPd:-KklStU09Q0ZapJt0NBI:bt1b19p) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer.go#L173)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

slightly scary that removing this line would still compile but would do the wrong thing
</blockquote></details>

Renamed.

---

*[pkg/util/tracing/tracer.go, line 221 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl4TEXbbdX7t0lZplh:-KklTapEgHTnHHedtu_9:b-e237nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer.go#L221)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can we unify this block with the other block above that calls `enableRecording`? So that the conditions for enabling recording are all visible in one place.

Also, it bothers me a little that a span now has two things that control recording now - the "snowball" baggage item and the `recordingGroup != nil`. Could we not use the baggage item any more, except when injecting the trace (so, for propagating the recording information across the network)? We'd terminate the baggage item in `JoinRemoteTrace`.
</blockquote></details>

Good idea, moved up.

We need to know whether we should put the snowball item at `Inject` time, so we need to propagate it. I think what you are suggesting is to set it at that time if the span is recording. I am not sure if we want to tie these things together. Right now it's entirely possible to record local spans without snowballing to remote spans (in fact we already do it for `debugTrace7881Enabled`).

---

*[pkg/util/tracing/tracer_span.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkj1BsQkvXsNm8f0De:-KklUlhFxEdpxmVkfGsE:b-usvj59) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

just wondering: why implement the interface on the value rather than pointer?
</blockquote></details>

It's a good question. The existing tracer implementations use this model so I stuck to that. Perhaps it makes it more clear that it's immutable.

---

*[pkg/util/tracing/tracer_span.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkko6jJGC3-4bjfKxXv:-KklVXy2M1RSvYfOeX2F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L74)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if this can be nil, please comment
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer_span.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkjEqI6sMH9-bhcddv:-KklVptry38XdpwVI5nC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L81)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: newline after?
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer_span.go, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkomp7XNpNug8r9eju:-KklWBXI2i1IpI79D3BI:b6uyz8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L89)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

but what if recording is started after the span is created (and presumably tags are usually set on the span when the span is created)?
</blockquote></details>

You only see the tags that get set after you start recording. Just like you only see the events that come after you start recording.  Can't say if that's the right thing but it's more efficient and we're not really using tags that much (in particular, we never set tags when we create the span). Added a TODO.

---

*[pkg/util/tracing/tracer_span.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkpFlbGAZo5M82WQQ7:-KklX4dayEpLK0I4EQ9k:b-uisiom) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L160)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: "to the record" is not very clear... Perhaps `ImportRemoteSpans()`? The `spanGroup` already talks about "spans from other hosts"
</blockquote></details>

Renamed. Note: we already have an `IngestRemoteSpans` but maybe that's fine? I think we will be consolidating these anyway when we switch to a proto for spans.

---

*[pkg/util/tracing/tracer_span.go, line 354 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkq3UgJNmQD1mn59sy:-KklYXvbOi2va8UJSwnD:b-92887b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L354)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think it'd be useful to keep track of which of the `spans/rawSpans` are finished, if we could get this information. I can add this later too, so I'd like to understand what's possible. I see that for `spans`, some of them are finished, some of them are not. And currently all the `rawSpans` are finished remote spans, right? But they don't have to be, if we'd have a way to ask remotes for unfinished spans.
</blockquote></details>

We can get this information from each span. We could easily change things so `GetRecording` doesn't set `Duration` for unfinished spans. I don't think if there is any other way to indicate an unfinished span in `RawSpan`.

---

*[pkg/util/tracing/tracer_span.go, line 78 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl-zZLzvq_yY5nrM3r:-KklVfYmQpa3LEcZ9jE9:blb4sfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L78)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think this one is unused (except for a place where it is always set to -1)
</blockquote></details>

It's used in getSpans. Added comment.

---

*[pkg/util/tracing/tracer_span.go, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkoom6L0KHW-DkdM5N:-KklYINbls3_zHL3vddT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L168)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

errors.New?
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer_span.go, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkkotCrFj2fcvOSRQVV:-KklYIiO7l4HeAR3fDBD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L177)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

might be good to expand on this. concretely, i have no idea what this means.
</blockquote></details>

Done.

---

*[pkg/util/tracing/tracer_span.go, line 263 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl8hrSBIx7Vy_zac-B:-KklYLBibqu4HrwZI5J0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer_span.go#L263)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Log*Fi*elds
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-:-KklZsDUv6EzrU7D9kze:b-hlc8bm)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 6 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/sql/distsqlrun/base.go, line 609 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KklauXghkhd3kyLzy0H:-KklauXghkhd3kyLzy0I:b-g1g284) ([raw file](https://github.com/cockroachdb/cockroach/blob/923acaed0591d75e99734bd99e038fb2bb259779/pkg/sql/distsqlrun/base.go#L609)):*
> req.TraceContext

I think this should have been `&traceContext`?

---

*[pkg/util/tracing/tracer_span.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkj1BsQkvXsNm8f0De:-KklbKay7eZoiPFIVBV6:b-jwk6bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L56)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's a good question. The existing tracer implementations use this model so I stuck to that. Perhaps it makes it more clear that it's immutable.
</blockquote></details>

Hm. I suggest we use a pointer - in truth, there's no difference except that pointers are more type-safe (because implementing an interface on a value implies that the pointer also implements it, but not vice versa)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 13 of 14 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/util/tracing/tracer.go, line 221 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkl4TEXbbdX7t0lZplh:-KklaSwaxuOCw5KtMETe:brgtg50) ([raw file](https://github.com/cockroachdb/cockroach/blob/82c9fda4a50350f63cdb6310a5513c8e92ba8d72/pkg/util/tracing/tracer.go#L221)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Good idea, moved up.

We need to know whether we should put the snowball item at `Inject` time, so we need to propagate it. I think what you are suggesting is to set it at that time if the span is recording. I am not sure if we want to tie these things together. Right now it's entirely possible to record local spans without snowballing to remote spans (in fact we already do it for `debugTrace7881Enabled`).
</blockquote></details>

hmmm ok if, it's possible to record without snowballing, then it makes sense

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 13 of 14 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/base.go, line 609 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KklauXghkhd3kyLzy0H:-KklidrHYctClc-VzqHZ:bueuwpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/923acaed0591d75e99734bd99e038fb2bb259779/pkg/sql/distsqlrun/base.go#L609)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> req.TraceContext

I think this should have been `&traceContext`?
</blockquote></details>

Lol, I guess I need some rest, thanks for keeping an eye out!

---

*[pkg/util/tracing/tracer_span.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-Kkkj1BsQkvXsNm8f0De:-Kklhk6NeQAGjsWlOd8E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed3c0e7a7c96130f14000b566b9b46f362aa2a41/pkg/util/tracing/tracer_span.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Hm. I suggest we use a pointer - in truth, there's no difference except that pointers are more type-safe (because implementing an interface on a value implies that the pointer also implements it, but not vice versa)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 5 files at r4.
Review status: 13 of 14 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-:-Kklm1GvHb1vx34OoKge:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I have made some additional changes discussed with @andreimatei offline: renamed `Force` to `Recordable` and changed the semantics so that, unless recording was actually started, child spans of `Recordable` spans are still `noop`. This will allow having a `Recordable` span for a SQL txn or session without foregoing the "noop" optimization for any child spans under that.

I will rerun all the manual testing.

---

Review status: 11 of 14 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-:-KkloLGgldRmgtKQkp10:bmtktnq)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 10 of 14 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/util/tracing/tracer_span.go, line 152 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-KkpyAqFNIAVNB7BQcxN:-KkpyAqG7hk5oHWtrrMN:b-szafvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c84a17c188e1a99bd5103f4159438d01723aec32/pkg/util/tracing/tracer_span.go#L152)):*
> ```Go
> 
> // GetRecording retrieves the current recording, if the span has
> // recording enabled.  This can be called while spans that are part of the
> ```

nit: double space

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16058#-:-KkpykPswdVyfkyGqJoC:bnfp4nl)*
<!-- Sent from Reviewable.io -->
